sirepo/package_data/static/js/sirepo.js

```diff
@@ -1038,6 +1041,345 @@ SIREPO.app.factory('timeService', function() {
+            [_e_needYes]: _s_status,
+            [_e_unloaded]: _s_initial,
+        },
+        [_s_ok]: {
+            [_e_authMissing]: _s_idle,
+            [_e_needYes]: _s_ok,
+            [_e_unloaded]: _s_initial,
+        },
+        [_s_status]: {
+            [_e_authError]: _s_idle,
+            [_e_authMissing]: _s_idle,
+            [_e_authSuccess]: _s_ok,
+            [_e_needYes]: _s_status,
+            [_e_unloaded]: _s_initial,
+        },
+    };
+
+    const _STATES = new Set(Object.keys(_TRANSITIONS));
+
+    const _EVENTS =  new Set(Object.values(_TRANSITIONS).map((x) => {return Object.keys(x);}).flat(1));
+
+    let _state = _s_initial;
+
+    const _STATE_QUERIES = {
+        isLoggedIn: (state) => {
+            return state === _s_ok;
+        },
+        showLogin: (state) => {
+            return [_s_creds, _s_idle].includes(state);
+        },
+        showLoginOrStatus: (state) => {
+            return [_s_auth, _s_creds, _s_idle, _s_status].includes(state);
+        },
+        showSbatchOptions: (state) => {
+            return state !== _s_notNeeded;
+        },
+    };
+
+    // This should be a static member, but that's not supported by jshint.
+    // So this code references private members
+    // See https://github.com/jshint/jshint/issues/3645
+    const _EVENT_QUERIES = {
+        hideCredsForm: (event) => {
+            return event._newState === _s_ok || event._event === _e_credsCancel;
+        },
+        isCredsFormBlank: (event) => {
+            return event._event === _e_loginClicked && event._oldState === _s_idle;
+        },
+        isCredsFormError: (event) => {
+            return [_e_authInvalid, _e_authError].includes(event._event) && event._newState === _s_creds;
+        },
+        isLoggedInFromCreds: (event) => {
+            return event._event === _e_authSuccess && event._oldState === _s_auth;
+        },
+        isLoginNotNeeded: (event) => {
+            return ! [_e_authSuccess, _e_needNo].includes(event._event);
+        },
+        requestSbatchLogin: (event) => {
+            return event._event === _e_credsConfirm && event._oldState === _s_creds;
+        },
+        requestSbatchLoginStatus: (event) => {
+            return event._oldState != _s_status && event._newState === _s_status;
+        },
+        showCredsForm: (event) => {
+            return event._event === _e_authMissing && [_s_ok, _s_status].includes(event._oldState);
+        },
+    };
+
+    class _Event {
+        constructor(event, arg) {
+            this._event = _assertEvent(event);
+            this._arg = this._assertArg(arg);
+            this._oldState = _state;
+            this._newState = this._nextState();
+        }
+
+        argProperty(name) {
+            if (name in this._arg) {
+                return this._arg[name];
+            }
+            throw new Error(`invalid arg property=${name} for event=${this._event} oldState=${this._oldState}`);
+        }
+
+        credsError() {
```

Comment:
I think the rule of 3 is important here. We have one case which is weird. We could add a function to service that accepts an event and returns the error message, but this has one use so it seemed like it was the simplest thing to do. If there were more methods like this or more uses of this method, it would make sense to generalize.